pkg/action/mount.go

```diff
@@ -192,68 +266,109 @@ func MountOverlayPath(cfg *v1.RunConfig, sysroot, overlayDir, path string) error
+			persistentVol = vol
 		}
 
-		data = data + fstab(part.Path, part.MountPoint, "auto", part.Flags)
+		data = data + fstab(vol.Device, vol.Mountpoint, "auto", vol.Options)
 	}
 
 	for _, rw := range spec.Ephemeral.Paths {
 		data += overlayLine(rw, constants.OverlayDir, constants.OverlayDir)
 	}
 
-	if ok, _ := elemental.IsMounted(cfg.Config, spec.Partitions.Persistent); ok {
+	if persistentVol != nil {
 		for _, path := range spec.Persistent.Paths {
 			if spec.Persistent.Mode == constants.OverlayMode {
-				data += overlayLine(path, constants.PersistentStateDir, constants.PersistentDir)
+				data += overlayLine(path, filepath.Join(persistentVol.Mountpoint, constants.PersistentStateDir), constants.PersistentDir)
 				continue
 			}
 
 			if spec.Persistent.Mode == constants.BindMode {
 				trimmed := strings.TrimPrefix(path, "/")
 				pathName := strings.ReplaceAll(trimmed, "/", "-") + ".bind"
-				stateDir := fmt.Sprintf("%s/%s", constants.PersistentStateDir, pathName)
+				stateDir := filepath.Join(persistentVol.Mountpoint, constants.PersistentStateDir, pathName)
 
 				data = data + fstab(stateDir, path, "none", []string{"defaults", "bind"})
 				continue
 			}
-
-			return fmt.Errorf("Unknown persistent mode '%s'", spec.Persistent.Mode)
+			errs = multierror.Append(errs, fmt.Errorf("Unknown persistent mode '%s'", spec.Persistent.Mode))
 		}
 	}
 
 	return cfg.Config.Fs.WriteFile(filepath.Join(spec.Sysroot, "/etc/fstab"), []byte(data), 0644)
 }
 
+func InitialFstabData(runner v1.Runner, sysroot string) (string, error) {
+	var data string
+
+	mounts, err := findmnt(runner, "/")
+	if err != nil {
+		return "", err
+	}
+	for _, mnt := range mounts {
+		if mnt.Mountpoint == sysroot {
+			data += fstab(mnt.Device, "/", "auto", mnt.Options)
+		} else if strings.HasPrefix(mnt.Mountpoint, sysroot) {
+			data += fstab(mnt.Device, strings.TrimPrefix(mnt.Mountpoint, sysroot), "auto", mnt.Options)
+		} else if strings.HasPrefix(mnt.Mountpoint, constants.RunElementalDir) {
+			data += fstab(mnt.Device, mnt.Mountpoint, "auto", mnt.Options)
+		} else if mnt.Mountpoint == constants.RunningStateDir {
+			data += fstab(mnt.Device, mnt.Mountpoint, "auto", mnt.Options)
+		}
+	}
+
+	return data, nil
+}
+
 func fstab(device, path, fstype string, flags []string) string {
 	if len(flags) == 0 {
 		flags = []string{"defaults"}
 	}
 	return fmt.Sprintf("%s\t%s\t%s\t%s\t0\t0\n", device, path, fstype, strings.Join(flags, ","))
 }
 
+func findmnt(runner v1.Runner, mountpoint string) ([]*v1.VolumeMount, error) {
+	mounts := []*v1.VolumeMount{}
+	output, err := runner.Run("findmnt", "-Rrfno", "SOURCE,TARGET,FSTYPE,OPTIONS", mountpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(strings.NewReader(strings.TrimSpace(string(output))))
+	for scanner.Scan() {
+		lineFields := strings.Fields(scanner.Text())
+		if len(lineFields) != 4 {
+			continue
+		}
+		if lineFields[2] == "btrfs" {
+			r := regexp.MustCompile(`(/.+)\[.*\]`)
```

Comment:
On btrfs mounted subvolume added in brackets something like `/dev/sda1[/@/.snapshots/1/snapshot]` hence we need to remove it  from the SOURCE field, that's why there this extra parsing logic here.